### PR TITLE
Teach naming conventions check to look for tests in src/main

### PR DESCRIPTION
**Pull request by nik9000**
Teaches the naming conventions check to fail the build when it finds
tests in `src/main/java`. Tests in `src/main/java` aren't run and that
is bad because tests that don't run provide a false sense of
security.

Edit: As it stands now this actually creates a new similar task to the
`namingConventions` task, `namingConventionsMain` that checks
the main classes of the `buildSrc` and `test/framework` projects,
looking for tests. This checking isn't done by the same task.

### CODE REVIEW

**Comment by jasontedor** [comment]
It looks like `LongGCDisruptionTest` is both incorrectly named (missing 's') and in the wrong place.

**Comment by nik9000** [comment]
> It looks like LongGCDisruptionTest is both incorrectly named (missing 's') and in the wrong place.

Want me to fix it in this PR or a followup?

**Comment by jasontedor** [comment]
> Want me to fix it in this PR or a followup?

This PR, and the additional check that you proposed via another channel.

**Comment by nik9000** [comment]
> This PR, and the additional check that you proposed via another channel.

I pushed another commit that should do that.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/NamingConventionsTask.groovy [hunk 1]
```diff
@@ -78,9 +68,19 @@ public class NamingConventionsTask extends LoggedExec {
         FileCollection extraClasspath = project.configurations.namingConventions
         dependsOn(extraClasspath)
 
+        FileCollection classpath = project.sourceSets.test.runtimeClasspath
+        inputs.files(classpath)
         description = "Runs NamingConventionsCheck on ${classpath}"
+        if (project.sourceSets.names.contains('main')) {
+            /* If there are main classes we'll have a look at them just to make
+             * sure there aren't file named like test classes in there too. */
+            inputs.files(project.sourceSets.main.runtimeClasspath)
```
Comment (rjernst) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/NamingConventionsTask.groovy]: I don't think this should be necessary? The test classpath should already include the main classpath.

**Review by rjernst — COMMENTED** [review]
This seems to complicate the naming conventions check, for an extreme edgecase. It is only possible that such an error would even compile if in the test framework, as LongGCDisruptionTest is.  I think it would be far simpler to have a separate task in test framework to look for such cases? I don't even know if it needs to be more than just looking at the classes that exist in test framework, and seeing if any end in Test, Tests, or IT.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/NamingConventionsTask.groovy (hunk 1 above)
Comment (nik9000) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/NamingConventionsTask.groovy]: Makes sense. I'll drop it.

**Comment by nik9000** [comment]
I think it is worth having the check. I don't think code review is a great mechanism for catching stuff like that. I didn't want to do anything just in test:framework because I though it'd be brittle to limit the check. And we already had the naming conventions check doing very similar things. I don't feel like looking at names is really good enough. I started out just looking at names and it missed the LongGCDisruptionTest.

All that being said, it is your build so if you want me to do something just for test:framework I'll change it. If you want me to drop the test changes and just move the classes I can do that too. Or some other thing I haven't thought of.

**Comment by rjernst** [comment]
> I don't think code review is a great mechanism for catching stuff like that

I don't understand this comment. Where did I suggest that? All I said was it is not possible to have a subclass of ESTestCase under main, except in test framework, so there is no use having this check anywhere else.

**Comment by nik9000** [comment]
> I don't understand this comment. Where did I suggest that? All I said was it is not possible to have a subclass of ESTestCase under main, except in test framework, so there is no use having this check anywhere else.

Sorry. Do you think it is worth adding something so this is only run on test:framework? Like a flag on the task or detecting that the main classes have junit as a dependency? Do you want a separate task entirely?

**Comment by rjernst** [comment]
I was suggesting an entirely separate task, in test:framework (which you already essentially have). I would not complicate the naming conventions task unless it is necessary, but I'm failing to see why it would be necessary.  I still think just checking names would be sufficient for this edge case. If you feel strongly, then I would say have a flag, and create a second task in test:framework which operates on the main classpath, and runs with this flag.

**Comment by nik9000** [comment]
@rjernst I liked the subclass checking a lot so I split the execution into two tasks, with the main class checking ones only being built on buildSrc and test:framework. You get the main class checking behavior by providing a flag to NamingConventionTask.

File: buildSrc/build.gradle [hunk 2]
```diff
@@ -156,4 +156,11 @@ if (project != rootProject) {
     testClass = 'org.elasticsearch.test.NamingConventionsCheckBadClasses$UnitTestCase'
     integTestClass = 'org.elasticsearch.test.NamingConventionsCheckBadClasses$IntegTestCase'
   }
+
+  task(namingConventionsMain, type: org.elasticsearch.gradle.precommit.NamingConventionsTask) {
```
Comment (rjernst) [inline buildSrc/build.gradle]: nit: can you use the style of task creation used throughout the rest of the build.gradle files, ie:
```
task namingConventionsMain(type: org.elasticsearch.gradle.precommit.NamingConventionsTask) {
```

File: test/framework/build.gradle [hunk 3]
```diff
@@ -63,3 +63,8 @@ thirdPartyAudit.excludes = [
   'org.easymock.IArgumentMatcher',
   'org.jmock.core.Constraint',
 ]
+
+task(namingConventionsMain, type: org.elasticsearch.gradle.precommit.NamingConventionsTask) {
```
Comment (rjernst) [inline test/framework/build.gradle]: ditto on nit about task creation syntax

**Review by rjernst — APPROVED** [review]
LGTM, one small nit.

File: buildSrc/build.gradle (hunk 2 above)
Comment (nik9000) [inline buildSrc/build.gradle]: Sure!

**Comment by nik9000** [comment]
Thanks for reviewing @rjernst!

**Pull request merged**